Update pipeline output targets validation with different writeMask

shrekshao opened the pull request:
Issue: #928

I believe the `attachment_compatibility.spec.ts` updates are already done in https://github.com/gpuweb/cts/pull/655
The missing part is validation at createRenderPipeline

Spec update: https://github.com/gpuweb/gpuweb/pull/1918
https://www.w3.org/TR/webgpu/#fragment-state

Failing current Chrome Canary, a fix is at https://dawn-review.googlesource.com/c/dawn/+/86980

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [ ] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [x] Tests are properly located in the test tree.
- [x] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [x] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

austinEng commented on src/webgpu/api/validation/createRenderPipeline.spec.ts:
```diff
@@ -422,24 +422,27 @@ g.test('pipeline_output_targets')
     u
       .combine('isAsync', [false, true])
+      .combine('writeMask', [0, 0xf])
       .combine('format', kRenderableColorTextureFormats)
```
see https://github.com/gpuweb/gpuweb/issues/2770. I think we should also test writeMasks that are not "all"

shrekshao commented:
Needs update based on (https://github.com/gpuweb/gpuweb/pull/1918#issuecomment-1113676136)

shrekshao commented:
Addressing https://github.com/gpuweb/gpuweb/pull/1918#issuecomment-1113676136
and https://dawn-review.googlesource.com/c/dawn/+/90040

Regroup getFragmentShaderCodeWithOutput into a helper to avoid duplicate code

austinEng commented on src/webgpu/api/validation/createRenderPipeline.spec.ts:
```diff
@@ -439,45 +395,72 @@ g.test('sample_count_must_be_valid')
       .combine('isAsync', [false, true])
-      .combine('format', kRenderableColorTextureFormats)
+      .combine('writeMask', [0, 0x1, 0x2, 0x4, 0x8, 0xf])
+      .combine('format', kRenderableColorTextureFormatsExtend)
```
only used once - maybe just do it inline?
```suggestion
      .combine('format',  [undefined, ...kRenderableColorTextureFormats] as const)
```

austinEng commented on src/webgpu/util/shader.ts:
```diff
@@ -0,0 +1,112 @@
+import { unreachable } from '../../common/util/util.js';
+
```
Please add this to docs/helper_index.txt

austinEng commented on src/webgpu/util/shader.ts:
```diff
@@ -0,0 +1,112 @@
+export function getFragmentShaderCodeWithOutput(
+  outputs: ({
+    values: readonly number[];
+    sampleType: GPUTextureSampleType;
```
odd that an input is GPUTextureSampleType when no sampling is giong on.

Perhaps it should take in `i32 | u32 | f32` and then have a separate helper that maps GPUTextureSampleType to that union

shrekshao commented:
Addressed review comments. (Better land after https://chromiumdash.appspot.com/commits?commit=c9ebe9104d79fdf8abf7b6f15e9056820e19f3ae&platform=Windows lands)
